### ISSUE TITLE
`from csv --separator ';'` doesnt' work

### ISSUE THREAD

**Issue by freijon** [nushell/nushell#10081]
### Describe the bug

When opening a csv file with a custom separator there are nonsensical error messages

### How to reproduce

Open a csv file with the following content:

```csv
column1;column2;column3
test1;test2;test3
test4;test5;test6
```
`open file.csv | from csv --separator ';'` produces:
```
Error: nu::shell::type_mismatch

  × Type mismatch.
   ╭─[entry #93:1:1]
 1 │ open test.csv | from csv --separator ';'
   · ──┬─
   ·   ╰── string
   ╰────
```

### Expected behavior

Should display as a table

### Configuration

| key                | value                                         |
| ------------------ | --------------------------------------------- |
| version            | 0.83.1                                        |
| branch             |                                               |
| commit_hash        |                                               |
| build_os           | linux-x86_64                                  |
| build_target       | x86_64-unknown-linux-gnu                      |
| rust_version       | rustc 1.71.1 (eb26296b5 2023-08-03) (gentoo)  |
| cargo_version      | cargo 1.71.1                                  |
| build_time         | 2023-08-17 07:15:22 +02:00                    |
| build_rust_channel | release                                       |
| allocator          | standard                                      |
| features           | dataframe, default, sqlite, trash, which, zip |
| installed_plugins  |                                               |

**Comment by lordkekz** [comment]
The unhelpful error occurs because the implicit `from csv` manages to interpret the almost-csv file as a regular csv. Thus, the error we get is actually about the explicit `from csv -s ';'` getting a list<string> instead of string and not about the forgotten `--raw`.
```nu
> open file.csv
╭───┬─────────────────────────╮
│ # │ column1;column2;column3 │
├───┼─────────────────────────┤
│ 0 │ test1;test2;test3       │
│ 1 │ test4;test5;test6       │
╰───┴─────────────────────────╯
```

After exploring the code a bit, I figured out that [FromCsv](https://github.com/nushell/nushell/blob/030e749fe742c84dfb603f74a2fc81803f26bd1e/crates/nu-command/src/formats/from/csv.rs#L172) just calls [from_delimited_data](https://github.com/nushell/nushell/blob/030e749fe742c84dfb603f74a2fc81803f26bd1e/crates/nu-command/src/formats/from/delimited.rs#L85) which in turn calls [collect_string_strict](https://github.com/nushell/nushell/blob/main/crates/nu-protocol/src/pipeline_data.rs#L360) in `PipelineData` which then raises the `ShellError::TypeMismatch` on incompatible values.

The `ShellError::TypeMismatch` is described as _"A command received an argument of the wrong type."_.
However, there also exists `ShellError::PipelineMismatch` and `ShellError::OnlySupportsThisInputType` which are both described as _"The pipelined input into a command was not of the expected type. For example, it might expect a string input, but received a table instead."_.
I think that since we are reading data from the input pipe and not from a regular argument, `PipelineMismatch` or `OnlySupportsThisInputType` might be a better fit than `TypeMismatch`. I prefer `OnlySupportsThisInputType` because it allows specifying the actually received input type, otherwise the error is still fairly unhelpful, since one might expect `open` to output a string. It also has a lot more usages according to my IDE.

I tried changing the error type to `OnlySupportsThisInputType` and providing the needed `Span`s for highlighting. When the pipeline input was a `Value` (if I understand correctly that means that it's fully in memory) it works perfectly:
```shell
Error: nu::shell::only_supports_this_input_type

  × Input type not supported.
   ╭─[entry #3:1:1]
 1 │  open file.csv | from csv -s ";"
   ·  ──┬─            ────┬───
   ·    │                 ╰── only string input data is supported
   ·    ╰── input type: table<column1;column2;column3: string>
   ╰────
```

However, `ListStream`s don't seem to have source `Span`s available, so for now I just create a span of the character before the `from csv`.
```shell
> def makestreamedoutput [] { seq 0 (2 ** 24) }
> makestreamedoutput | from csv -s ";"

Error: nu::shell::only_supports_this_input_type

  × Input type not supported.
   ╭─[entry #8:1:1]
 1 │ makestreamedoutput | from csv -s ";"
   ·                     ┬────┬───
   ·                     │    ╰── only string input data is supported
   ·                     ╰── input type: list
   ╰────
```
- Why do ListStreams not have a span?
- Is is safe to assume that `ListStreams` always represent Nu's `list<any>` datatype (since we probably don't yet know the types of all the items)?

I also considered providing a hint specific to this case, but it's difficult/impossible to catch it at parse time (we don't know what open will produce) and also not easily possible for either of the commands to tell if it was used in this way. Overall I think it's not worth the difficulty and the above error messages along with reading `help open` or `help from csv` will make it clear.

Do you think that solves the issue? I'll open a PR if you like it.

**Comment by amtoine** [comment]
@lordkekz 
i'd say just saying the separator is invalid would be enough, right?
maybe something like
```nushell
> open test.csv | from csv --separator ';'
Error: nu::shell::type_mismatch

  × Type mismatch.
   ╭─[entry #93:1:1]
 1 │ open test.csv | from csv --separator ';'
   · ──┬─
   ·   ╰── expected a "Comma Separated Value" file, found separator ";"
   ╰────
```

**Comment by lordkekz** [comment]
I agree, that message would point to the problem much quicker. But technically, the CSV is valid. I'm worried that `open` (or the internally used `from csv`) shouldn't fail with an error in that case, especially because this unhelpful error message only appears when the given CSV is valid (but accidentally so).
I think that `open test.csv` on its own should work.
Since we discover the output type of `open` only after parsing the file, the error should be raised when `from csv` reads the pipe. 

Ideally, I'd like to "enrich" the error message with hints based on the larger context. For example:
```
Error: nu::shell::only_supports_this_input_type

  × Input type not supported.
   ╭─[entry #3:1:1]
 1 │  open file.csv | from csv -s ";"
   ·  ──┬─            ────┬───
   ·    │                 ╰── only string input data is supported
   ·    ╰── input type: table<column1;column2;column3: string>
   ·    
   ·    Hint: You piped the output of 'open' into 'from csv', did you mean 'open --raw'?
   ╰────
```
Right now, this doesn't seem to be easily possible, because the parameters of the error are set when it's raised inside the function, which doesn't know about the pipeline at large. Maybe we could add a step for this hint enrichment before printing the error?

**Comment by amtoine** [comment]
@lordkekz 
if i'm not mistaken, the following data is not valid CSV :thinking: 
```csv
column1;column2;column3
test1;test2;test3
test4;test5;test6
```

it's more like an "SSV" for Semicolon Separated Values, but i'm not shocked `open test.csv` doesn't like it.

as @fdncred said above, when a file has an extension which is not 100% correct regarding the actual content of the file, you should use something like
```nushell
open --raw test.csv | from csv --separator ";"
```

as for the error enrichment, i think better errors is always good if it does not make them too verbose :+1:

**Comment by amtoine** [comment]
hope i'm not missing something obvious about the content of `test.csv` but i'm having trouble seeing it as valid CSV :confused:

**Comment by fdncred** [comment]
ya, semicolon separate csv is not a csv, nor is it a ssv which is space separated values.

**Comment by amtoine** [comment]
> ya, semicolon separate csv is not a csv, nor is it a ssv which is space separated values.
>
yeah i made up the "SSV" name above, SSV is with spaces :+1:

**Comment by lordkekz** [comment]
@amtoine Well, it's not meant as a CSV. But if you just blindly follow the CSV syntax as described in [RFC 4180 section 2](https://datatracker.ietf.org/doc/html/rfc4180#section-2), you can interpret it as a CSV consisting of only one column (and the rows are the lines of the input).  And `open test.csv` just interprets it as you would if you assumed it was intended as a CSV. Just because it _could_ be semicolon-separated doesn't mean that it can't be understood as comma separated. This is the result we get when we apply the RFC's syntax on the example:
```nu
> open file.csv
╭───┬─────────────────────────╮
│ # │ column1;column2;column3 │
├───┼─────────────────────────┤
│ 0 │ test1;test2;test3       │
│ 1 │ test4;test5;test6       │
╰───┴─────────────────────────╯
```

So, we shouldn't reject one-column CSVs or CSVs which incidentally contain other potential delimiters. Recognizing the details (e.g. delimiter) of a non-standard file variant should be up to the user IMO. In the initial post, OP did already recognize that; the error message should instead point towards the missing `--raw`.

**Comment by amtoine** [comment]
@lordkekz 
ooh i see :scream: 
yeah i could be interpreted as a CSV with a single column :+1:
